File: tks/ApplicationCore/DomainModel/src/main/kotlin/model/values/AccessionNumber.java
package model.values;

import exceptions.TypeValidationFailedException;
import lombok.Value;

@Value
public class AccessionNumber {
    String value;

    public AccessionNumber(String value) throws TypeValidationFailedException {
        if(!value.matches(format)) throw new TypeValidationFailedException();
        this.value = value;
    }

    static final String format = "\\b[A-Z]{4}-[0-9]{3}\\b";
}

File: tks/ApplicationCore/DomainModel/src/main/kotlin/domain/model/values/AccessionNumber.kt
package domain.model.values

import domain.exceptions.TypeValidationFailedException

@JvmInline
value class AccessionNumber(val value: String) {
    companion object {
        val format = Regex("\\b[A-Z]{4}-[0-9]{3}\\b")
    }

    init {
        if (!value.matches(format)) throw TypeValidationFailedException()
    }
}